### ISSUE TITLE
Different performance when training with single GPU vs. multiple GPUs

### ISSUE THREAD

**Issue by rahuln** [huggingface/accelerate#90]
I'm currently using `accelerate` to fine-tune a huggingface pretrained Transformer with some additional classification heads, and I'm finding that performance when using multiple GPUs is much worse than with a single GPU, even when using the same batch size, learning rate, and number of training steps/epochs. I'm using accelerate to parallelize the training loop over multiple GPUs, but the validation/test set evaluation is a custom function that isn't easily adapted to use with accelerate so I'm doing that part on a single GPU in the main process. To run the entire script on a single GPU vs. multiple GPUs, I just adjust the `--num_processes` argument for `accelerate launch` as well as the batch size to match, for example:

`accelerate launch --num_processes 1 <script> --batch_size 32` (for 1 GPU)

`accelerate launch --num_processes 4 <script> --batch_size 8` (for 4 GPUs)

The multi-GPU training seems to be running fine, in the sense that running `nvidia-smi` shows all 4 GPUs being fully utilized and the training data loader is the correct length for the given batch size (same length for both of the commands above), but there's still a drop in performance in the multi-GPU case. When printing output on each GPU, the processes do seem to be waiting for the main process to finish running the evaluation function as expected. This also doesn't seem to just be an issue of running single-GPU evaluation within a multi-GPU training loop, since loading the saved model weights after training and re-running the evaluation on a single GPU gives the same performance.

Any help is appreciated, thanks!

Pseudocode:
```python
# set up accelerator
ddp_kwargs = DistributedDataParallelKwargs(find_unused_parameters=True)
accelerator = Accelerator(kwargs_handlers=[ddp_kwargs])

# set up device (for evaluation)
device = accelerator.device

model = ... # initialize model
optimizer = ... # initialize optimizer
loader = ... # initialize training data loader

valid_dataset = ... # initialize validation dataset
test_dataset = ... # initialize test dataset

# prepare model, optimizer, data loader
model, optimizer, loader = accelerator.prepare(model, optimizer, loader)

# training loop
for epoch in range(epochs):
    model.train()
    for inputs, targets in loader:
        outputs = model(inputs)
        loss = loss_function(outputs, targets)
        optimizer.zero_grad()
        accelerator.backward(loss)
        optimizer.step()

    # evaluate on validation set with unwrapped model in main process (single-GPU)
    if accelerator.is_main_process:
        unwrapped_model = accelerator.unwrap_model(model).to(device)
        unwrapped_model.eval()
        metrics = calculate_metrics(unwrapped_model, valid_dataset, device)
        print(metrics)

# evaluate on test set with unwrapped model in main process (single-GPU)
if accelerator.is_main_process:
    unwrapped_model = accelerator.unwrap_model(model).to(device)
    unwrapped_model.eval()
    metrics = calculate_metrics(unwrapped_model, test_dataset, device)
    print(metrics)
```

Version information:
- `torch`: 1.6.0
- `transformers`: 3.3.1
- `accelerate`: 0.3.0
- CUDA: 10.1

**Comment by sgugger** [comment]
I'm afraid we can't help very much without understanding your training code more. First, when you mention performance, do you mean you have a lower metric value with distributed training or that the time spent is not satisfactory?

There is a generic test in the CI that training yields the same results on one or two GPUs so you should get the same results in theory. In practice, it does depend on the randomness you have so knowing what the model is and how the data is processed would be useful. All in all, to investigate more and help, I would need some reproducible code.

**Comment by rahuln** [comment]
Thanks for the prompt reply! By "performance" I mean a lower metric value, the training time seems to be as expected (faster with multiple GPUs). The model is essentially just a pretrained Transformer with multiple output heads, I've included some pseudocode below (unfortunately, it won't be easy to share the full codebase and data to reproduce the experiment).

The data does actually include some randomness: each data example is the text of a head and tail entity for a link in a knowledge graph, with additional examples where the head/tail entity is sampled randomly to generate negatives for a max-margin ranking loss. I'm not handling the randomness in any specific way other than setting
```python
np.random.seed(seed)
torch.manual_seed(seed)
torch.cuda.manual_seed_all(seed)
```
at the beginning of the training script – is there something more that needs to be done with the `accelerator` to properly handle the random state for multiple GPUs?

I also just wanted to make sure that: (1) performing single-GPU evaluation within a multi-GPU training loop as I've written in the code above should work fine (in theory); and (2) that I'm not missing any key component beyond `accelerator.prepare` and `accelerator.backward` (what I have is fine as far as I can tell from the `accelerate` docs).

Model pseudocode:
```python
class MyModel(torch.nn.Module):

    def __init__(self, model_name, num_classes):
        super(MyModel, self).__init__()
        self.encoder = AutoModel.from_pretrained(model_name)
        hidden_size = self.encoder.config.hidden_size
        self.linear1 = torch.nn.Linear(hidden_size, 1)
        self.linear2 = torch.nn.Linear(hidden_size, num_classes)

    def forward(self, inputs):
        outputs = self.encoder(input_ids=inputs.input_ids,
                               token_type_ids=inputs.token_type_ids,
                               attention_mask=inputs.attention_mask,
                               return_dict=True)
        cls_token = outputs.last_hidden_state[:, 0, :]
        outputs = {1 : self.linear1(cls_token), 2 : self.linear2(cls_token)}
        return outputs
```

**Comment by sgugger** [comment]
Your usage of the library seemed fine, I don't think there is a problem there. The randomization seems more likely the cause: by setting the seeds like this you essentially will get the same randomness on all GPUs, which means that each of the four processes will perform the same "data augmentation" (I'm not sure exactly if that is the right term in your use case but you get the idea) on different samples (when just one GPU will do a different data augmentation on each sample).

Could you try without setting any seed and see where it leads you?

**Comment by rahuln** [comment]
Not setting any seed worked, thanks! I'm now getting similar performance using either a single GPU or 4 GPUs.

Is there an appropriate way to handle setting a random seed when using `accelerate`? It would be nice to have some level of reproducibility without having the previous issue of the same randomness on all GPUs.

**Comment by sgugger** [comment]
For reproducibility, you can still set a seed, you should just set a different one for each process. One way to do this easily with Accelerate is:
```
np.random.seed(seed + accelerator.process_index)
torch.manual_seed(seed + accelerator.process_index)
torch.cuda.manual_seed_all(seed + accelerator.process_index)
```

**Comment by rahuln** [comment]
Got it! I think this solves the problem I was having, closing the issue now – thanks for your help!
